NPE when querying `_all` field with no fields

DELETE my_index

    POST my_index/my_type/1
    {}
    
    GET my_index/_search
    {
      "query": {
        "match": {
          "_all": "foo"
        }
      }
    }

Returns an NPE with the following stack trace:

    Failed to execute [org.elasticsearch.action.search.SearchRequest@402fa749]
    RemoteTransportException[[Havok][inet[/127.0.0.1:9300]][indices:data/read/search[phase/query]]]; nested: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: ElasticsearchException; nested: NullPointerException;
    Caused by: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: ElasticsearchException; nested: NullPointerException;
        at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:157)
        at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:334)
        at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:346)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
        at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:340)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:722)
    Caused by: ElasticsearchException; nested: NullPointerException;
        at org.elasticsearch.ExceptionsHelper.convertToElastic(ExceptionsHelper.java:57)
        at org.elasticsearch.search.internal.ContextIndexSearcher.createNormalizedWeight(ContextIndexSearcher.java:133)
        at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:485)
        at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:173)
        at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:694)
        at org.apache.lucene.search.IndexSearcher.searchAfter(IndexSearcher.java:410)
        at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:439)
        at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:151)
        ... 9 more
    Caused by: java.lang.NullPointerException
        at org.elasticsearch.common.lucene.all.AllTermQuery.rewrite(AllTermQuery.java:130)
        at org.apache.lucene.search.IndexSearcher.rewrite(IndexSearcher.java:786)
        at org.apache.lucene.search.IndexSearcher.createNormalizedWeight(IndexSearcher.java:833)
        at org.elasticsearch.search.internal.ContextIndexSearcher.createNormalizedWeight(ContextIndexSearcher.java:130)
        ... 15 more

--- a/core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java
+++ b/core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java
@@ -25,6 +25,7 @@ import org.apache.lucene.index.PostingsEnum;
 import org.apache.lucene.index.Term;
 import org.apache.lucene.index.Terms;
 import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchNoDocsQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.TermQuery;
 import org.apache.lucene.search.payloads.AveragePayloadFunction;
@@ -124,13 +125,20 @@ public final class AllTermQuery extends PayloadTermQuery {
 
     @Override
     public Query rewrite(IndexReader reader) throws IOException {
+        boolean fieldExists = false;
         boolean hasPayloads = false;
         for (LeafReaderContext context : reader.leaves()) {
             final Terms terms = context.reader().terms(term.field());
-            if (terms.hasPayloads()) {
-                hasPayloads = true;
-                break;
+            if (terms != null) {
+                fieldExists = true;
+                if (terms.hasPayloads()) {
+                    hasPayloads = true;
+                    break;
+                }
             }
+        }
+        if (fieldExists == false) {
+            return new MatchNoDocsQuery();
         }
         if (hasPayloads == false) {
             TermQuery rewritten = new TermQuery(term);

--- a/core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java
+++ b/core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java
@@ -138,7 +138,9 @@ public final class AllTermQuery extends PayloadTermQuery {
             }
         }
         if (fieldExists == false) {
-            return new MatchNoDocsQuery();
+            Query rewritten = new MatchNoDocsQuery();
+            rewritten.setBoost(getBoost());
+            return rewritten;
         }
         if (hasPayloads == false) {
             TermQuery rewritten = new TermQuery(term);

--- a/core/src/test/java/org/elasticsearch/search/query/SearchQueryTests.java
+++ b/core/src/test/java/org/elasticsearch/search/query/SearchQueryTests.java
@@ -1868,6 +1868,24 @@ public class SearchQueryTests extends ElasticsearchIntegrationTest {
     }
 
     @Test
+    public void testAllFieldEmptyMapping() throws Exception {
+        client().prepareIndex("myindex", "mytype").setId("1").setSource("{}").setRefresh(true).get();
+        SearchResponse response = client().prepareSearch("myindex").setQuery(matchQuery("_all", "foo")).get();
+        assertNoFailures(response);
+    }
+
+    @Test
+    public void testAllDisabledButQueried() throws Exception {
+        createIndex("myindex");
+        assertAcked(client().admin().indices().preparePutMapping("myindex").setType("mytype").setSource(
+                jsonBuilder().startObject().startObject("mytype").startObject("_all").field("enabled", false)));
+        client().prepareIndex("myindex", "mytype").setId("1").setSource("bar", "foo").setRefresh(true).get();
+        SearchResponse response = client().prepareSearch("myindex").setQuery(matchQuery("_all", "foo")).get();
+        assertNoFailures(response);
+        assertHitCount(response, 0);
+    }
+
+    @Test
     public void testIndicesQuery() throws Exception {
         createIndex("index1", "index2", "index3");
 

